iceoryx2-cal/src/shm_allocator/mod.rs

```diff
@@ -13,44 +13,147 @@
 pub mod bump_allocator;
 pub mod pool_allocator;
 
-use std::{alloc::Layout, ptr::NonNull};
+use std::{alloc::Layout, fmt::Debug, ptr::NonNull};
 
 pub use iceoryx2_bb_elementary::allocator::AllocationError;
 use iceoryx2_bb_elementary::{allocator::BaseAllocator, enum_gen};
 
-pub trait ShmAllocatorConfig: Copy + Default {}
+/// Trait that identifies a configuration of a [`ShmAllocator`].
+pub trait ShmAllocatorConfig: Copy + Default + Debug {}
 
-#[derive(Clone, Copy, Eq, PartialEq, Debug)]
-pub struct PointerOffset(usize);
+/// Defines the [`SegmentId`] of a [`SharedMemory`](crate::shared_memory::SharedMemory)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SegmentId(u8);
 
-impl PointerOffset {
-    pub fn new(value: usize) -> PointerOffset {
+impl SegmentId {
+    /// Creates a new [`SegmentId`] from a given value.
+    pub fn new(value: u8) -> Self {
         Self(value)
     }
 
-    pub fn value(&self) -> usize {
+    /// Returns the underlying value of the [`SegmentId`]
+    pub fn value(&self) -> u8 {
         self.0
     }
+
+    /// Returns the maximum value the [`SegmentId`] supports.
+    pub const fn max_segment_id() -> u8 {
+        u8::MAX
+    }
 }
 
-enum_gen! { ShmAllocationError
+const NUMBER_OF_BITS_IN_BYTE: usize = 8;
+
+/// An offset to a [`SharedMemory`](crate::shared_memory::SharedMemory) address. It requires the
+/// [`SharedMemory::payload_start_address()`](crate::shared_memory::SharedMemory::payload_start_address())
+/// of the corresponding [`SharedMemory`](crate::shared_memory::SharedMemory) to be converted into
+/// an actual pointer.
+///
+/// Contains the offset and the corresponding [`SegmentId`].
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub struct PointerOffset(u64);
+
+impl PointerOffset {
+    /// Creates a new [`PointerOffset`] from the given offset value with the [`SegmentId`] == 0.
+    pub fn new(offset: usize) -> PointerOffset {
+        const SEGMENT_ID: u64 = 0;
+        Self(
+            (offset as u64) << (core::mem::size_of::<SegmentId>() * NUMBER_OF_BITS_IN_BYTE)
+                | SEGMENT_ID,
```

Comment:
```suggestion
            (offset as u64) << SegmentIdUnderlyingType::BITS | SEGMENT_ID,
```

Reply:
Wouldn't be your suggestion a bit more cumbersome and error prone?!

This would require us to add this const type, write tests for it and verify that the `BITS` value fits?